fix: Left-to-right text alignment is visible in the email body for RTL languages

retrogtx opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Tried to fix #16681 (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

retrogtx committed:
Add RTL styling to BaseEmailHtml component and RTL class conditionally in BaseScheduledEmail

retrogtx committed:
add email head back

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=16697) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=16697) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=16697) it.</sub>

anikdhabal commented on packages/emails/src/components/WhenInfo.tsx:
```diff
@@ -53,8 +53,10 @@ export function WhenInfo(props: {
     attendee: props.calEvent.attendees[0],
   });
 
+  const isRTL = locale === "he";
+
```
Hebru is not the only RTL language; there are more.

retrogtx replied:
sure thing! will look up and add another commit to this soon. just need to update an array value I guess.

retrogtx committed:
added arabic

sean-brydon reviewed (changes requested):
Lets do this using INTL 
```
const ar = new Intl.Locale("ar");
console.log(ar.getTextInfo()); // { direction: "rtl" }
console.log(ar.getTextInfo().direction); // "rtl"
```

retrogtx commented:
> Lets do this using INTL
> 
> ```
> const ar = new Intl.Locale("ar");
> console.log(ar.getTextInfo()); // { direction: "rtl" }
> console.log(ar.getTextInfo().direction); // "rtl"
> ```

so should I update the `packages/lib/i18n.ts` file and hardcode he and ae there, then call the locales?

retrogtx commented:
> > Lets do this using INTL
> > ```
> > const ar = new Intl.Locale("ar");
> > console.log(ar.getTextInfo()); // { direction: "rtl" }
> > console.log(ar.getTextInfo().direction); // "rtl"
> > ```
> 
> so should I update the `packages/lib/i18n.ts` file and hardcode he and ae there, then call the locales?

i did do it, i hope this TODO goal in `WhenInfo.tsx` isn't interfering with this possibly
`// TODO: Use browser locale, implement Intl in Dayjs maybe?`

I modified i18n.ts as follows:
const rtlLocales = ["ar", "he"];

export function getTextDirection(locale: string): "ltr" | "rtl" {
  return rtlLocales.includes(locale) || rtlLocales.includes(locale.split("-")[0]) ? "rtl" : "ltr";
}

export function isRTL(locale: string): boolean {
  return getTextDirection(locale) === "rtl";
}

retrogtx committed:
add rtl logic to i18n.ts

Amit91848 commented on packages/emails/src/components/WhenInfo.tsx:
```diff
@@ -53,8 +54,11 @@ export function WhenInfo(props: {
   });
 
+  const textDirection = getTextDirection(props.locale);
+  const isRTL = textDirection === "rtl";
```
Since you have isRtl function wouldn’t it be better to just use that directly?

Amit91848 commented on packages/emails/src/templates/BaseScheduledEmail.tsx:
```diff
@@ -29,6 +29,7 @@ export const BaseScheduledEmail = (
   } & Partial<React.ComponentProps<typeof BaseEmailHtml>>
 ) => {
   const { t, timeZone, locale, timeFormat: timeFormat_ } = props;
+  const isRTL = locale === "he" || locale === "ar";
```
Same here directly call the isRTL function

Amit91848 commented on packages/emails/src/components/BaseEmailHtml.tsx:
```diff
@@ -24,78 +25,137 @@ export const BaseEmailHtml = (props: {
+  locale: string;
 }) => {
+  const textDirection = getTextDirection(props.locale);
+  const isRTL = textDirection === "rtl";
```
And here as well

retrogtx replied:
yes sir that's what I did, I wrote it that way

but this was suggested 

https://github.com/calcom/cal.com/pull/16697#pullrequestreview-2312131712

and I think it'll be better to keep everything in one place, incase this needs to be updated we might need to update it only once instead of three files. 

what do both of you think?

Amit91848 replied:
Yes that's fine. I meant calling the isRTL function directly since it calls getTextDirection and matches if it is "rtl", instead of you calling getTextDirection here and then manually matching it everywhere

https://github.com/calcom/cal.com/blob/e6f1cd84a9b7f8874758e512c66ea914eb207b1a/packages/lib/i18n.ts#L23-L25

```suggestion
  const isRightToLeft = isRTL(props.locale);
```

Amit91848 replied:
https://github.com/calcom/cal.com/pull/16697#pullrequestreview-2312131712  Also this comment means you should use `Intl.Locale` and fetch information about its alignment instead of us maintaining a list, would be easier in case we add more languages.

Amit91848 commented on packages/lib/i18n.ts:
```diff
@@ -12,3 +12,14 @@ export const localeOptions = locales.map((locale) => ({
+  return rtlLocales.includes(locale) || rtlLocales.includes(locale.split("-")[0]) ? "rtl" : "ltr";
+}
+
+export function isRTL(locale: string): boolean {
```
Also coming to think of it, isRTL is a very good function name, you can't tell what it does. Please consider renaming it to something more descriptive.

```suggestion
export function isLocaleRightToLeft(locale: string): boolean {
```

Or something

retrogtx replied:
I'll consider it haha, I just thought it would make sense to name it as rtl for "right to left" 

rn I'm away from home for a while so can't access the laptop but I'll try to do it soon

Amit91848 replied:
Yeah no worries. 
It makes sense now because we know the context. Someone not knowing would have difficulty understanding what `isRTL` means

anikdhabal commented on packages/lib/i18n.ts:
```diff
@@ -12,3 +12,14 @@ export const localeOptions = locales.map((locale) => ({
 export const defaultLocaleOption = localeOptions.find(
   (locale) => locale.value === i18n.defaultLocale
 ) as (typeof localeOptions)[number];
+
+// List of known RTL locales, it returnbs rtl or ltr for respective locales
+const rtlLocales = ["ar", "he"];
+
+export function getTextDirection(locale: string): "ltr" | "rtl" {
+  return rtlLocales.includes(locale) || rtlLocales.includes(locale.split("-")[0]) ? "rtl" : "ltr";
```
Why are we doing extra things? In the future, if we add any new RTL locale, we will need to update it. Also, since we have access to the new intl object, why not use it? Sean also mentioned it.
Read more about it:- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/getTextInfo#examples

anikdhabal commented on packages/emails/src/components/BaseEmailHtml.tsx:
```diff
@@ -24,78 +25,137 @@ export const BaseEmailHtml = (props: {
   hideLogo?: boolean;
+  locale: string;
 }) => {
+  const textDirection = getTextDirection(props.locale);
+  const isRTL = textDirection === "rtl";
```
You can simply get the text direction by passing it to new `Intl.Locale(props.locale)` and extracting it using the `getTextInfo()` method

retrogtx replied:
hi, i did implement this

is a try catch block (because of browsers that dont support the Intl library) good enough?

 try {
    if (typeof Intl.Locale !== "undefined" && "getTextInfo" in Intl.Locale.prototype) {
      const extendedLocale = new Intl.Locale(locale) as ExtendedLocale;
      if (extendedLocale.getTextInfo) {
        return extendedLocale.getTextInfo().direction;
      }
    }
    // Fallback for browsers that don't support Intl.Locale.prototype.getTextInfo
    const rtlLocales = ["ar", "he", "fa", "ur"];
    return rtlLocales.some((lang) => locale.toLowerCase().startsWith(lang)) ? "rtl" : "ltr";
  } catch (error) {
    console.error(`Error determining text direction for locale ${locale}:`, error);
    return "ltr";
  }
  
  
  lmk if this looks good, have to fix some type errors in this

anikdhabal reviewed (changes requested):
Left comment

retrogtx committed:
use Intl with fallback for old browsers

Amit91848 commented:
Weird does .getTextInfo() even work in any browser? I keep getting it is not a function in edge, brave and chrome 🤔

Amit91848 reviewed (changes requested):
Type checks are failing @retrogtx , can you please fix those

retrogtx commented:
> Type checks are failing @retrogtx , can you please fix those

yes I'm trying to fix, my laptop is lagging locally too so its taking a bit time to see changes being reflected too

retrogtx committed:
fix type errors

retrogtx commented:
> [anikdhabal](/anikdhabal)

does the core logic look good now @anikdhabal ?

retrogtx committed:
fix type for feedback argument

retrogtx committed:
fix: Fix type for feedback, add _replacements

retrogtx commented:
I think it works, but need to test the functionality again.
There is still a "PR Update / required (pull_request_target) " failing. I think I took care of types in the past 2 commits.

Any idea why this is happening?

Amit91848 commented on packages/lib/i18n.ts:
```diff
@@ -12,3 +12,28 @@ export const localeOptions = locales.map((locale) => ({
   (locale) => locale.value === i18n.defaultLocale
 ) as (typeof localeOptions)[number];
+
+interface ExtendedLocale extends Intl.Locale {
+  getTextInfo?: () => { direction: "ltr" | "rtl" };
+}
```
I don't see getTextInfo working anywhere, it's just mentioned in the docs so it's best to just remove this and use .textInfo.direction which works.

Amit91848 commented on packages/lib/i18n.ts:
```diff
@@ -12,3 +12,28 @@ export const localeOptions = locales.map((locale) => ({
+
+export function getTextDirection(locale: string): "ltr" | "rtl" {
+  try {
+    if (typeof Intl.Locale !== "undefined" && "getTextInfo" in Intl.Locale.prototype) {
+      const extendedLocale = new Intl.Locale(locale) as ExtendedLocale;
+      if (extendedLocale.getTextInfo) {
+        return extendedLocale.getTextInfo().direction;
+      }
+    }
+    // Fallback for gbrowsers that don't support Intl.Locale.prototype.getTextInfo
+    const rtlLocales = ["ar", "he"];
+    return rtlLocales.some((lang) => locale.toLowerCase().startsWith(lang)) ? "rtl" : "ltr";
```
```suggestion
    const locale = new Intl.Locale(locale);
    if (locale.textInfo.direction) rerturn locale.textInfo.direction;
    // fallback
    const rtlLocales = ["ar", "he"];
    return rtlLocales.some((lang) => locale.toLowerCase().startsWith(lang)) ? "rtl" : "ltr";
```

Do something like this and rewrite this

retrogtx replied:
hi again, I made a [commit](https://github.com/calcom/cal.com/pull/16697/commits/feb168b5d9edbe6b0ed66034170cb13f3bc536c8) that solves this

Amit91848 reviewed (changes requested):
Left some suggestions @retrogtx

retrogtx committed:
use textInfo.direction
